app/bundles/AssetBundle/Tests/EventListener/ReportSubscriberTest.php

```diff
@@ -122,23 +156,63 @@ public function __construct()
     private function createTranslatorMock(): TranslatorInterface
     {
         return new class() implements TranslatorInterface {
-            public function trans($id, array $parameters = [], $domain = null, $locale = null)
+            /**
+             * @param array<int|string> $parameters
+             */
+            public function trans($id, array $parameters = [], $domain = null, $locale = null): string
             {
                 return '[trans]'.$id.'[/trans]';
             }
 
+            /**
+             * @param array<int|string> $parameters
+             *
+             * @return string
+             */
             public function transChoice($id, $number, array $parameters = [], $domain = null, $locale = null)
             {
                 return '[trans]'.$id.'[/trans]';
             }
 
-            public function setLocale($locale)
+            public function setLocale($locale): void
             {
             }
 
             public function getLocale()
             {
+                return '';
             }
         };
     }
+
+    public function testGroupByIfNotConfigured(): void
+    {
+        $report             = new Report();
+        $report->setSource(ReportSubscriber::CONTEXT_ASSET_DOWNLOAD);
+        $event              = new ReportGeneratorEvent($report, [], $this->queryBuilder, $this->channelListHelper);
+        $subscriber         = new ReportSubscriber($this->companyReportData, $this->downloadRepository);
+        $this->queryBuilder->method('from')->willReturn($this->queryBuilder);
+
+        $this->queryBuilder->expects($this->once())
+            ->method('groupBy')
+            ->with('ad.asset_id');
+
+        $subscriber->onReportGenerate($event);
+    }
+
+    public function testGroupByIfConfigured(): void
+    {
+        $report             = new Report();
+        $report->setSource(ReportSubscriber::CONTEXT_ASSET_DOWNLOAD);
+        $this->queryBuilder->method('from')->willReturn($this->queryBuilder);
+        $report->setGroupBy(['a.id' => 'desc']);
+        $event              = new ReportGeneratorEvent($report, [], $this->queryBuilder, $this->channelListHelper);
+        $subscriber         = new ReportSubscriber($this->companyReportData, $this->downloadRepository);
+
+        $this->queryBuilder->expects($this->never())
+            ->method('groupBy')
+            ->with('ad.asset_id');
```

Comment:
How come it's never called? Shouldn't it be grouping by a.id like this?
```

```
```suggestion
        $this->queryBuilder->expects($this->once())
            ->method('groupBy')
            ->with('a.id');
```

Reply:
This expects require defined some methods inside reporbuilder and quieryBuilder. In this workflow groupBy is not defined and I added just tests for hasGroupBy. It is enought?

Reply:
Gotcha. Looks good!